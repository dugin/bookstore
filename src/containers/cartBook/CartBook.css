.CartBook__item {
  border: 1px solid #d1d1d1;
  margin-top: -1px;
  background-color: #fff;
  display: flex;
  position: relative;
  align-items: center; }
  .CartBook__item__price {
    font-size: 13px;
    font-weight: bold;
    display: flex;
    flex-direction: column;
    padding-left: 0 !important;
    padding-right: 5px !important;
    align-items: flex-end; }
    @media (min-width: 768px) {
      .CartBook__item__price {
        margin-top: 20px;
        margin-left: -15%;
        font-size: 16px; } }
    .CartBook__item__price__amount, .CartBook__item__price__unit {
      font-size: 11px;
      color: #939393; }
  .CartBook__item__remove {
    margin-top: 20px;
    font-size: 12px;
    display: flex;
    align-items: center;
    justify-content: flex-end; }
    @media (max-width: 767px) {
      .CartBook__item__remove {
        bottom: 10px;
        right: 5px;
        position: absolute !important; } }
    .CartBook__item__remove .material-icons {
      font-size: 22px;
      margin-right: 3px; }
  .CartBook__item .Book__actions__quantityContainer {
    position: absolute;
    left: 110px;
    bottom: 10px; }
  .CartBook__item .Book__wrapper {
    display: flex;
    padding: 10px 0; }
    .CartBook__item .Book__wrapper .Book__figureContainer__img {
      width: 80px;
      margin-right: 15px; }
    .CartBook__item .Book__wrapper .Book__figureContainer__caption__name {
      margin-top: 0;
      font-size: 14px; }
    .CartBook__item .Book__wrapper .Book__figureContainer__caption__subtitle {
      font-size: 10px; }
    .CartBook__item .Book__wrapper .Book__figureContainer__caption__description {
      display: none; }
