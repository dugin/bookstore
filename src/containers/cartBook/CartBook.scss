@import '../../variables';

.CartBook {

  &__item {
    border: 1px solid $color__lineCard--gray;
    margin-top: -1px;
    background-color: $color--white;
    display: flex;
    position: relative;
    align-items: center;


    &__price {

      font-size: 13px;
      font-weight: bold;
      display: flex;
      flex-direction: column;
      padding-left: 0 !important;
      padding-right: 5px !important;
      align-items:flex-end ;

      @media   (min-width: $screen-sm-min){
        margin-top: 20px;
        margin-left: -15%;
        font-size: 16px;

      }

      &__amount, &__unit{
        font-size: 11px;
        color: $color__text--gray;
      }
    }

    &__remove {
      margin-top: 20px;
      font-size: 12px;
      display: flex;
      align-items: center;
      justify-content: flex-end;


      @media   (max-width: $screen-xs-max){
        bottom: 10px;
        right: 5px;
        position: absolute !important;
      }

      .material-icons {
        font-size: 22px;
        margin-right: 3px;
      }
    }
    .Book__actions__quantityContainer {
      position: absolute;
      left: 110px;
      bottom: 10px;
    }

    .Book__wrapper {
      display: flex;
      padding: 10px 0;

      .Book__figureContainer {

        &__img {
          width: 80px;
          margin-right: 15px;

        }

        &__caption {

          &__name {
            margin-top: 0;
            font-size: 14px;
          }

          &__subtitle {
            font-size: 10px;
          }

          &__description {
            display: none;
          }

        }
      }
    }
  }

}