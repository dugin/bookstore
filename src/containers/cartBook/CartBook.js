import React, {Component} from 'react';
import BookInfo from "../../components/bookInfo/BookInfo";
import {connect} from 'react-redux';
import './CartBook.css';
import QuantityButtons from "../quantityButtons/QuantityButtons";
import * as utils from "../../utils/booksUtils";
import NumberFormat from 'react-number-format';
import {removeFromCart, addToCart} from "../../actions/cart";


export class CartBook extends Component {

    setPrice = (b) => {
        return utils.setPrice(b.discount, b.price);
    };


    onAddOrRemove = (quantity, b) => {
        const diff = quantity - b.amount;
        this.props.changeAmount(b.book, diff);
    };

    removeBook = (id) => {
        this.props.removeFromCart(id);
    };

    render() {

        return (
            <div className="CartBook ">
                {this.props.books.map(b => (
                    <div className="CartBook__item row" key={b.book.id}>
                        <div className="col-sm-7 col-md-8 col-9 pr-0">
                            <BookInfo   {...b}/>
                            <QuantityButtons
                                amount={b.amount}
                                onAddOrRemove={(quantity) => this.onAddOrRemove(quantity, b)}
                            />
                        </div>
                        <div className="CartBook__item__price col-sm-5 col-md-3 col-3 ">
                            <div>
                                <span className="CartBook__item__price__amount mr-1"> {b.amount}x</span>
                                <NumberFormat className="  CartBook__item__price__unit"
                                              value={this.setPrice(b.book)}
                                              decimalPrecision={2} displayType={'text'} decimalSeparator=','
                                              prefix={'R$ '}/>
                            </div>
                            <NumberFormat className=" CartBook__item__price__total"
                                          value={b.amount * this.setPrice(b.book)}
                                          decimalPrecision={2} displayType={'text'}
                                          decimalSeparator=',' prefix={'R$ '}/>
                        </div>

                        <a onClick={() =>this.removeBook(b.book.id)} className="CartBook__item__remove  pr-0 col-md-2">
                            <i  className="material-icons">highlight_off</i>excluir</a>
                    </div>
                ))}

            </div>
        )
    }

}

export const mapDispatchToProps = (dispatch) => {
    return {
        changeAmount: (book, amount) => dispatch(addToCart(book, amount)),
        removeFromCart: (id) => dispatch(removeFromCart(id)),
    }
};


export const mapStateToProps = (state, props) => ({books: state.books});


export default connect(mapStateToProps, mapDispatchToProps)(CartBook);