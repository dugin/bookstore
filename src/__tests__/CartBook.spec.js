import React from 'react';
import {shallow,} from 'enzyme';
import toJson from 'enzyme-to-json';
import {CartBook, mapDispatchToProps, mapStateToProps} from "../containers/cartBook/CartBook";
import {props} from '../__mocks__/bookMock';

describe('Book', () => {

    delete props.addToCart;
    const books = [props];

    const cartBook = shallow(<CartBook removeFromCart={jest.fn()} changeAmount={jest.fn()}  books={books}/>);

    it('renders properly', () => {
        expect(toJson(cartBook)).toMatchSnapshot();
    });

    it('mapStateToProps properly', () => {
        expect(mapStateToProps({books: [props]}, null)).toEqual({books: cartBook.instance().props.books});
    });

    it('clicks on `excluir` button, triggers `removeFromCart`', () => {
        const btn = cartBook.find('.CartBook__item__remove');

        btn.simulate('click');

        expect(cartBook.instance().props.removeFromCart).toHaveBeenCalled();
    });

    it('clicks on add or remove buttons, triggers `changeAmount`', () => {

        cartBook.instance().onAddOrRemove(5, {amount: 1});

        expect(cartBook.instance().props.changeAmount).toHaveBeenCalled();
    });

    it('mapDispatchToProps properly', () => {
        expect(JSON.stringify(mapDispatchToProps(jest.fn()))).toEqual(JSON.stringify({changeAmount: jest.fn(), removeFromCart: jest.fn()}))
    })



});
