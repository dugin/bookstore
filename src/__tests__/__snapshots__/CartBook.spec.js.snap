// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book renders properly 1`] = `
<div
  className="CartBook "
>
  <div
    className="CartBook__item row"
    key="nggnmAEACAAJ"
  >
    <div
      className="col-sm-7 col-md-8 col-9 pr-0"
    >
      <BookInfo
        amount={2}
        book={
          Object {
            "description": "You've experienced the shiny, point-and-click surface of your Linux computer—now dive below and explore its depths with the power of the command line. The Linux Command Line takes you from your very first terminal keystrokes to writing full programs in Bash, the most popular Linux shell. Along the way you'll learn the timeless skills handed down by generations of gray-bearded, mouse-shunning gurus: file navigation, environment configuration, command chaining, pattern matching with regular expressions, and more. In addition to that practical knowledge, author William Shotts reveals the philosophy behind these tools and the rich heritage that your desktop Linux machine has inherited from Unix supercomputers of yore. As you make your way through the book's short, easily-digestible chapters, you'll learn how to: * Create and delete files, directories, and symlinks * Administer your system, including networking, package installation, and process management * Use standard input and output, redirection, and pipelines * Edit files with Vi, the world’s most popular text editor * Write shell scripts to automate common or boring tasks * Slice and dice text files with cut, paste, grep, patch, and sed Once you overcome your initial \\"shell shock,\\" you'll find that the command line is a natural and expressive way to communicate with your computer. Just don't be surprised if your mouse starts to gather dust. A featured resource in the Linux Foundation's \\"Evolution of a SysAdmin\\"",
            "discount": 5,
            "id": "nggnmAEACAAJ",
            "imageLinks": Object {
              "smallThumbnail": "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api",
              "thumbnail": "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api",
            },
            "price": 45,
            "title": "The Linux Command Line",
          }
        }
      />
      <QuantityButtons
        amount={2}
        onAddOrRemove={[Function]}
      />
    </div>
    <div
      className="CartBook__item__price col-sm-5 col-md-3 col-3 "
    >
      <div>
        <span
          className="CartBook__item__price__amount mr-1"
        >
           
          2
          x
        </span>
        <NumberFormat
          allowNegative={true}
          className="  CartBook__item__price__unit"
          decimalPrecision={2}
          decimalSeparator=","
          displayType="text"
          isAllowed={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          onMouseUp={[Function]}
          prefix="R$ "
          suffix=""
          type="text"
          value={40}
        />
      </div>
      <NumberFormat
        allowNegative={true}
        className=" CartBook__item__price__total"
        decimalPrecision={2}
        decimalSeparator=","
        displayType="text"
        isAllowed={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        onMouseUp={[Function]}
        prefix="R$ "
        suffix=""
        type="text"
        value={80}
      />
    </div>
    <a
      className="CartBook__item__remove  pr-0 col-md-2"
      onClick={[Function]}
    >
      <i
        className="material-icons"
      >
        highlight_off
      </i>
      excluir
    </a>
  </div>
</div>
`;
